src/Compilers/Core/Portable/PEWriter/BlobBuilder.cs

```diff
@@ -677,15 +679,27 @@ public void WriteBytes(Stream source, int byteCount)
 
             int bytesToCurrent = Math.Min(FreeBytes, byteCount);
 
-            int bytesRead = source.Read(_buffer, Length, bytesToCurrent);
+            int bytesRead = source.TryReadAll(_buffer, Length, bytesToCurrent);
             AddLength(bytesRead);
 
+            if (bytesRead != bytesToCurrent)
+            {
+                return bytesRead;
+            }
+
             int remaining = byteCount - bytesToCurrent;
-            if (remaining > 0 && bytesRead == bytesToCurrent)
+            if (remaining > 0)
             {
                 Expand(remaining);
-                AddLength(source.Read(_buffer, 0, remaining));
+                bytesRead = source.TryReadAll(_buffer, 0, remaining);
+                AddLength(bytesRead);
+                
+                if (bytesRead != remaining)
```

Comment:
I am not sure it is useful having this ```if```. Instead, we could simply say ```bytesRead += bytesToCurrent;```  and do ```return bytesRead;``` rather than ```return byteCount;``` at the end of the function.